UI-Testing

williamw04 opened the pull request:
# Description  
To improve site stability and catch potential issues, our team (@aravindb24, @shankarp05, and I) worked on adding end-to-end (e2e) UI testing to the test suite. Testing the layout, dynamic values, and page behavior helps ensure that contributions do not unintentionally break the site.  

We developed a [UI Testing Documentation](https://github.com/OpenEnergyDashboard/DesignDocs/blob/main/uiTesting/UITesting.md) to provide guidance and address common issues, making it easier for others to learn and gain hands-on experience with UI testing.  

For UI and UX testing, we leveraged **Cypress** as our primary framework. To ensure consistent test results across environments, we containerized the testing environment using **Docker**.  

### Areas for Future Improvement  
There are still areas that need further attention, such as testing Redux state. Please refer to the [Limitations section](https://github.com/OpenEnergyDashboard/DesignDocs/blob/main/uiTesting/UITesting.md) of the documentation for more details.  

### Testing Progress  

**Shankar**  
- **hideOptions** – Testing options dropdown functionality  
- **initLangTest** – Language feature (Spanish only)  
- **pagesGroups** – Group page dropdown functionality  
- **pagesMeters** – Meters dropdown functionality  
- **tooltipIcon** – Tooltip attribute test  

**Aravind**  
- **general_ui** – General homepage testing (visibility of elements and layout)  

**William**  
- **line** – Testing line graph meters and group combinations (in progress)  

### Additional Files  
- **docker-compose.yml**  
- **cypress.config.ts** – Configures Cypress (test locations, support files, screenshot directories)  
- **support/commands.ts** – (Currently not utilized)  
- **support/e2e.ts** – Imports commands.js  

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

There are TODOs in the code and the design document about remaining items. This is a demonstration of UI testing and not complete.

aravindb24 committed:
fixed cypress testing: to perform ui-testing run docker compose --profile ui-testing up

aravindb24 committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into development

aravindb24 committed:
small change to docker-compose to autostart cypress when host is ready

aravindb24 committed:
restored comments

shankarp05 committed:
added language test

aravindb24 committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into Aravind

aravindb24 committed:
ss

aravindb24 committed:
attach shell to test now, and run cypress commands in shell, git will not track screenshots

aravindb24 committed:
Merge pull request #1 from aravindb24/Aravind

Basic Setup for Testing

williamw04 committed:
moved cypress into src and updated cypress config

aravindb24 committed:
Initial Test for buttons

aravindb24 committed:
Intiial form tests

aravindb24 committed:
Intial dropdown test

aravindb24 committed:
Intial test for links

aravindb24 committed:
Initial tests for modals

aravindb24 committed:
Initial tests for table data

aravindb24 committed:
Intial test for checkboxes

aravindb24 committed:
Intial interactive test (radio added)

aravindb24 committed:
Initial test for dynamic elements

aravindb24 committed:
delete duplicate test and unexpose postgres serv

williamw04 committed:
creating test for line menu option: including line option selected, checked groups so far

shankarp05 committed:
beginning test for hide options button

shankarp05 committed:
adding to hide options button test

shankarp05 committed:
finishing hide options test

shankarp05 committed:
starting tests for testing pages dropdown

testing groups and meters links from pages dropdown

shankarp05 committed:
adding to groups test

shankarp05 committed:
finishing group button testing

shankarp05 committed:
starting meters button testing

shankarp05 committed:
finishing meters test

cy.contains("Meters").click() did not work although it did for Groups. used selector instead.

shankarp05 committed:
beginning tooltip icon testing

shankarp05 committed:
adding to tooltip icon test

shankarp05 committed:
finishing tooltip test

shankarp05 committed:
Merge pull request #2 from aravindb24/shankar

Shankar

williamw04 committed:
Merge branch 'Aravind' into development

williamw04 committed:
Merge branch 'williamw04_ui' into development

williamw04 committed:
relocated testing file location

williamw04 committed:
line ui testing expanded and notes regarding future testing

williamw04 committed:
Merge branch 'development' into development

williamw04 committed:
pushing changes to fix check

huss commented on .gitignore:
```diff
@@ -55,4 +55,7 @@ nohup.out
\ No newline at end of file
+src/server/tmp/*
+
+#ui-testing screenshots
```
This folder is no longer present nor does OED normally have a folder for images in the code. Can this be removed or is it needed because the express.config.ts file puts screenshots there?

williamw04 replied:
So cypress will automatically put screenshots when there's an error in the test. This was configured to save to the path     screenshotsFolder: 'cypress/screenshots/e2e', within cypress.config.ts

I think that means if the folder isn't included within the development repo, it will automatically create one when needed so the folder and the images don't have to be included.

huss replied:
The cypress/ directory is not mapped from the container back to the developer's machine so it does not live there. However, I have sometimes seen it show up in VSC and the file system while the cypress container is running (not sure why). Thus, to avoid it ever showing up in git, I modified the directory in .gitignore to avoid seeing them.

huss commented on cypress.config.ts:
```diff
@@ -0,0 +1,16 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+ 
```
Extra spaces at start of the next two lines. Also, using spaces not tabs for indenting.

williamw04 replied:
Are we not suppose to use spaces for indenting?

huss replied:
I systematically converted the spaces to tabs. The developer first steps (https://openenergydashboard.org/developer/gettingStarted/#ide) does mention this. I also autoformatted in VSC to get rid of extra spaces.

huss commented on src/cypress/e2e/general_ui.cy.ts:
```diff
@@ -0,0 +1,96 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+ 
+describe('UI Functionality Tests for Open Energy Dashboard', () => {
```
It looks like the .ts files use spaces not tabs for indenting. Could it be changed (VSC can do this)?

huss replied:
See other comment for my change.

huss commented on src/cypress/e2e/initLangTest.cy.ts:
```diff
@@ -0,0 +1,25 @@
+
+  it("should update the UI and React state when the language is changed", () => {
+    // Open the language selection dropdown
+
```
Extra blank line.

huss commented on src/cypress/e2e/line.cy.ts:
```diff
@@ -0,0 +1,85 @@
+		cy.visit('/');
+	});
+
+  // Graph Type is Line
```
This is the default case but the initial type of graphic can be changed on the admin settings page. Maybe add a comment that test should be enhanced?

huss replied:
I added a TODO about this.

huss commented on src/cypress/e2e/line.cy.ts:
```diff
@@ -0,0 +1,85 @@
+
+  });
+  // Checking all group options
+  it('groups should be clickable and display 10 options and 1 incompatiable option', () => {
```
I think this assumes the default developer data is loaded without changes. Maybe worth noting.

williamw04 replied:
That is true where do you think is best to note this down? Should is be apart of the main test?

huss replied:
I added this to the design doc.

huss commented on src/cypress/e2e/line.cy.ts:
```diff
@@ -0,0 +1,85 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+describe('template spec', () => {
+	beforeEach(() => {
+		// Visit the OED application
```
This describe text is also used below. Should it be unique or something else?

williamw04 replied:
So the describe block is borrowed from Mocha's test infereceinterface. Describe explains the test within the block and can be used to describe different test within a file. So in the file I named it template spec since this test is a good test file for people to refer to when learning about testing with cypress.

huss commented on src/cypress/e2e/test.cy.ts:
```diff
@@ -0,0 +1,70 @@
+		cy.visit("/");
+	});
+
+	// it('should display the "Graph Type" button and toggle the dropdown', () => {
```
Since this is all commented out, what is it for?

huss reviewed (changes requested):
Thanks to @williamw04, @aravindb24 & @shankarp05 for this PR. I've made a few comments to consider but nothing substantial. One question is that when I tried to use this I got an error:
```
docker compose --profile ui-testing up
I got the error ``dependency failed to start: container oed-web-1 has no healthcheck configured``.
```
Does this happen for you? I wanted to try it and update the design doc with any changes for this PR so am curious how to get it to run properly.

williamw04 commented:
> Thanks to @williamw04, @aravindb24 & @shankarp05 for this PR. I've made a few comments to consider but nothing substantial. One question is that when I tried to use this I got an error:
> 
> ```
> docker compose --profile ui-testing up
> I got the error ``dependency failed to start: container oed-web-1 has no health check configured``.
> ```
> 
> Does this happen for you? I wanted to try it and update the design doc with any changes for this PR so am curious how to get it to run properly.

The error arises from there beingno health check defined in the web service. 

In the cypress service section, a "depends on web" is defined means the container depends/relies on the web to be running properly (have a healthy status so the healthcheck is the dependency that should have been started but was not sorry error). This is checked through health checks which was not previously defined. So moving forward if you were to define depend on you would have to have a defined health check in the service that is being depended upon.

Now with this being said the web service depends on the database so a health check must also be defined in the service.

The issue is health checks would happen in running OED normally (docker compose up) which might be an undesired feature. The solution is to remove the health checks.

williamw04 committed:
fix to health issue and addressed other comments

huss committed:
formatting

- Convert space indenting to tabs.
- Autoformat to fix up.

huss committed:
location of screenshots

- Modify .gitignore so actual location of cyprus screenshots.
- Remove unneeded training comma.

huss commented:
> The issue is health checks would happen in running OED normally (docker compose up) which might be an undesired feature. The solution is to remove the health checks.

I've tried and so far there is no issue. My recollection is that in the past the database check did not actually verify that the DB could accept request and only that it is running. The install script tries to check/correct for this. I've made myself a action item to look into this more in the future but it seems fine unless someone finds an issue.

huss commented:
Thanks to @williamw04 for the updates. Given them and my changes it is looking good.

I do have an issue where several of the tests are failing. I've attached the terminal output. Do you ever see this? Any thoughts? I did try logging in as admin outside the testing (could only do if not running the special cypress startup) to see if that might be the cause but it did not seem to help.

[cyprusFailures.txt](https://github.com/user-attachments/files/18389991/cyprusFailures.txt)

huss committed:
TODO for default line graphic

williamw04 commented:
> Thanks to @williamw04 for the updates. Given them and my changes it is looking good.
> 
> I do have an issue where several of the tests are failing. I've attached the terminal output. Do you ever see this? Any thoughts? I did try logging in as admin outside the testing (could only do if not running the special cypress startup) to see if that might be the cause but it did not seem to help.
> 
> [cyprusFailures.txt](https://github.com/user-attachments/files/18389991/cyprusFailures.txt)

For the general_ui.cy aravind said he was unable to get most of his test working.
For the line.cy test i think it only passes if you install the test data.

williamw04 commented:
> > The issue is health checks would happen in running OED normally (docker compose up) which might be an undesired feature. The solution is to remove the health checks.
> 
> I've tried and so far there is no issue. My recollection is that in the past the database check did not actually verify that the DB could accept request and only that it is running. The install script tries to check/correct for this. I've made myself a action item to look into this more in the future but it seems fine unless someone finds an issue.

I wasn't aware that there was already a database check prior, so with this in mind we could remove all the health checks and depends on from the docker yml file

williamw04 committed:
adjusted the screenshots path in cypress config, removed eextra line from initLangTest and changed description in describe routine

williamw04 committed:
resolved merge conflicts

williamw04 committed:
updated git ignore to ignore screenshots within the cypress screenshots directory

williamw04 committed:
commented out general_ui test needs to be fixed in the future

huss committed:
some changes

- General tests now left one that worked.
- Add TODOs for items.
- fix spaces vs tabs for intending.
- Misc edits.

huss commented:
> > > The issue is health checks would happen in running OED normally (docker compose up) which might be an undesired feature. The solution is to remove the health checks.
> > 
> > 
> > I've tried and so far there is no issue. My recollection is that in the past the database check did not actually verify that the DB could accept request and only that it is running. The install script tries to check/correct for this. I've made myself a action item to look into this more in the future but it seems fine unless someone finds an issue.
> 
> I wasn't aware that there was already a database check prior, so with this in mind we could remove all the health checks and depends on from the docker yml file

We decided to leave the check. It may be possible to remove the loop in the install script but that is future work.

huss commented:
> > Thanks to @williamw04 for the updates. Given them and my changes it is looking good.
> > I do have an issue where several of the tests are failing. I've attached the terminal output. Do you ever see this? Any thoughts? I did try logging in as admin outside the testing (could only do if not running the special cypress startup) to see if that might be the cause but it did not seem to help.
> > [cyprusFailures.txt](https://github.com/user-attachments/files/18389991/cyprusFailures.txt)
> 
> For the general_ui.cy aravind said he was unable to get most of his test working. For the line.cy test i think it only passes if you install the test data.

I've added a TODO and info in the design document about needing a clean DB. I also commented out the tests that were failing and left the one that works.

huss reviewed (approved):
Thanks to @williamw04 for continuing to work on this and resolving the comments. It is now working as desired at this time.
